design/velero-upgrade.md

```diff
@@ -0,0 +1,80 @@
+- Update all CRDs by running ```velero install --crds-only --dry-run -o yaml | kubectl apply -f -```
+- Update Velero server and daemonset image by running ```kubectl set image```  
+
+For a complete example of the existing upgrade approach, refer to the Velero document at https://velero.io/docs/v1.9/upgrade-to-1.9/.
+
+The manual approach works for the existing scenarios, however, if we want to do some complex tasks during upgrade (i.e., installing a new module or changing/adding complex parameters), it will be very unfriendly in the user experiences.  
+For an concrete example, in v1.10, we have done some refactors to Velero's pod volume backup (a.k.a Restic backup formerly), two of the refactors are:  
+- Rename some Velero server and daemonset parameters
+- Rename Velero daemonset from "restic" to "node-agent"
+
+As a result, the existing upgrade approach is not sufficient:
+- If we still use manual steps, we have to tell users the detailed parameters that have been renamed, and users then needs to check their existence and modify them one by one.
+- If we still use manual steps, users have to delete the old restic daemonset and run ```velero install``` with the full parameters they specified during the fresh installation, because for one thing, Kubernetes doesn't support to rename a resource; for another, Velero doesn't provide a way to install a single resource only. 
+
+## Solution
+
+We want to add a new command besides ```velero install``` called ```velero upgrade```. In a upgrade case, users could call ```velero upgrade``` without bothering to know the details. This bring below benefits:  
+- Solve the upgrade problems in v1.10
+- Make Velero more user friendly, because the upgrade process is significantly simplified
+- Support future evolvements. For example, in future, Velero will have various data movement features, for which we will add new moudles inevitably. Then we can add enhancement incrementally to ```velero upgrade```, because inside the command, we can do whatever complex things 
+
+The ```velero upgrade``` command will fully replace the existing manual steps in the aforementioned Velero document.
+
+## Design
+
+At present, ```velero upgrade``` will do below things:
+- Get the new version from the client binary itself, that is, the upgrade target version
+- Get the old version from the existing velero server, that is, the current version
+- Retrieve Velero server deployment object
+- Retrieve Velero daemonset object
+- Iterate from current version through target version, apply the changes (i.e., the image, name and parameter changes) to the deployment object once per version. For example, when upgrading from 1.9 to 1.11, we need to apply the changes between 1.9 and 1.10 first, then apply the changes between 1.10 and 1.11. See the details below for how to drive the changes apply
```

Comment:
Do we need to wait for the v1.10 to be running before proceeding to v1.11? If the CRD of v1.10 is not applied at this point of time, there may be risk that the v1.10 server fail to be running.

Reply:
Good point, let me modify the order of the steps to solve this problem.

Reply:
Changed the order of the action steps.